File: cerebras/test/elu.py
import torch, pytest
import custom_module

@pytest.mark.parametrize("input, alpha", [
    (torch.tensor([3.0, 0.0, -2.0]), 1.0),   
    (torch.tensor([5.0]), -3.0), 
    (torch.tensor([1.5, -1.5]), -5.5),  
    (torch.tensor([0.0]), 2.5),  
])
def test_elu(input, alpha):

    # Create an instance of Heaviside
    ch = custom_module.ELU()

    # Perform custom Heaviside
    custom_elu = ch.compute_elu(input, alpha)

    torch_elu = torch.nn.functional.elu(input, alpha)

    # Assert that the custom result matches the expected output
    assert torch.allclose(custom_elu, torch_elu, rtol=1e-05, atol=1e-05), \
    f"custom_elu({input}, {alpha}) != torch.elu({input}, {alpha})"

File: cerebras/src/elu.cpp
#include <elu.h>

namespace custom_namespace {
  // Custom ELU function implementation
  torch::Tensor ELU::compute_elu(const torch::Tensor& input, double alpha) {
      // Initialize the result tensor
      torch::Tensor result = torch::zeros_like(input);
  
      // For x >= 0, ELU(x) = x
      result = torch::where(input >= 0, input, result);
  
      // For x < 0, ELU(x) = alpha * (exp(x) - 1)
      result = torch::where(input < 0, alpha * (torch::exp(input) - 1), result);
  
      return result;
  }
}